### PR TITLE
feat: improve serialization for errors and bigints

### PR DESCRIPTION
**Pull request by alexander-akait**
feat:
- support `case` for errors (we will use it in future to provide original error from loader)
- support bigint

### <samp>🤖 Generated by Copilot at b1dac04</samp>

This pull request adds `bigint` support to the serialization system, which allows webpack to handle large integer values in its data structures. It modifies the `BinaryMiddleware`, `ObjectMiddleware`, `ErrorObjectSerializer`, and `types.js` modules to handle the `bigint` type, and adds the `parserOptions` property to the `.eslintrc.js` file to enable linting for ECMAScript 2020. It also adds a test case to the `Compiler-filesystem-caching.test.js` module to verify the serialization and deserialization of `bigint` values.

### <samp>🤖 Generated by Copilot at b1dac04</samp>

*  Add `bigint` support to the serialization system ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5R112-R114), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R87), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R335-R344), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R861-R879), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-a7a17d233126cd636c2f953911efa1c4872c84fe7255cab8878b1237a180c5e9R400-R402), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-4c5780a135e9a531231c1d6b54849b7c0efbfa6373a8b760cbeb0b6bff233da0L9-R9))
  * Define the header byte for `bigint` values in `BinaryMiddleware.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R87))
  * Handle the conversion of `bigint` values to and from binary buffers in `BinaryMiddleware.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R335-R344), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-276033679497d08f13669fab20cfcd1de13ec0bb36792d71404a15ab59627f29R861-R879))
  * Handle the serialization of objects with circular references that contain `bigint` values in `ObjectMiddleware.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-a7a17d233126cd636c2f953911efa1c4872c84fe7255cab8878b1237a180c5e9R400-R402))
  * Add `bigint` to the type definition of `PrimitiveSerializableType` in `types.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-4c5780a135e9a531231c1d6b54849b7c0efbfa6373a8b760cbeb0b6bff233da0L9-R9))
  * Add a test case for `bigint` values and edge cases in `Compiler-filesystem-caching.test.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-35817f6d3958df306aef88203a15f24dd7348e0ea8993740d329bd4f068b5b1eR44-R118))
* Fix the type error for error objects with a custom `cause` property in `ErrorObjectSerializer.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-6ee65ea18a780a266c1e0a9525ad2d975b8db9cbb2bc7a1c07130861672aba8cR24), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-6ee65ea18a780a266c1e0a9525ad2d975b8db9cbb2bc7a1c07130861672aba8cR35-R36))
  * Add type casts to the `write` and `read` methods of `ErrorObjectSerializer.js` to avoid the type error ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-6ee65ea18a780a266c1e0a9525ad2d975b8db9cbb2bc7a1c07130861672aba8cR24), [link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-6ee65ea18a780a266c1e0a9525ad2d975b8db9cbb2bc7a1c07130861672aba8cR35-R36))
* Remove the console log statement in `ObjectMiddleware.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-a7a17d233126cd636c2f953911efa1c4872c84fe7255cab8878b1237a180c5e9R692-R693))
  * Delete the debugging statement in the `deserialize` method of `ObjectMiddleware.js` ([link](https://github.com/webpack/webpack/pull/17282/files?diff=unified&w=0#diff-a7a17d233126cd636c2f953911efa1c4872c84fe7255cab8878b1237a180c5e9R692-R693))

### CODE REVIEW

File: lib/serialization/BinaryMiddleware.js [hunk 1]
```diff
@@ -331,6 +332,16 @@ class BinaryMiddleware extends SerializerMiddleware {
...
+				case "bigint": {
+					const value = thing.toString();
+					const len = Buffer.byteLength(value);
+					allocate(len + HEADER_SIZE + I32_SIZE);
+					writeU8(BIGINT_HEADER);
+					writeU32(len);
+					currentBuffer.write(value, currentPosition);
+					currentPosition += len;
+					break;
+				}
```
Comment (alexander-akait) [inline lib/serialization/BinaryMiddleware.js:344]: Why it is storing as string:
- we can't use `writeBigInt64LE`, because BigInt can be longer than 64bits
- storing by bytes (uint8 or other) will require multiple/shift them in deserialization
- we can use `thing.toString(16)` to reduce size, but I doesn't bring a lot of wins, `4294967296n.toString(16)` -> `100000000`, also it will require to store sing (-/+) and will multiple on -1

so storing them as string, I think, is the best solution

Comment (alexander-akait) [reply lib/serialization/BinaryMiddleware.js:402]: Improved, now we store small bigints as int8 (`n <= BigInt(127) && n >= BigInt(-128)`) and int32 (`n <= BigInt(2147483647) && n >= BigInt(-2147483648)`) as regular numbers

File: .eslintrc.js [hunk 2]
```diff
@@ -109,6 +109,9 @@ module.exports = {
 			env: {
 				"jest/globals": true
 			},
+			parserOptions: {
+				ecmaVersion: 2020
+			},
```
Comment (alexander-akait) [inline .eslintrc.js:114]: Support `12n` BigInt notation in tests

File: lib/serialization/BinaryMiddleware.js [hunk 3]
```diff
@@ -331,6 +353,62 @@ class BinaryMiddleware extends SerializerMiddleware {
...
+								);
+								currentPosition += I8_SIZE;
+								n--;
+								i++;
+							}
+							i--;
+							break;
+						}
+						case 1: {
+							let n = 1;
```
Comment (alexander-akait) [inline lib/serialization/BinaryMiddleware.js:384]: In theory it can be optimized and we can write sections, but it requires a lot of work and our serialization optimized for numbers and strings (they are used more often than BigInts), also it will cause a big cache broken errors on CI (i.e. if you have cache from old webpack version and the new version of webpack can't deserialize because we will use new bytes for headers)

**Review by TheLarkInn — APPROVED** [review]
This is great! Nice cache size performance optimization.
